src/opentelemetry.ts

```diff
@@ -0,0 +1,246 @@
+import { IORedisInstrumentation } from '@opentelemetry/instrumentation-ioredis';
+import { PgInstrumentation } from '@opentelemetry/instrumentation-pg';
+import { FastifyInstrumentation } from '@opentelemetry/instrumentation-fastify';
+import { GraphQLInstrumentation } from '@opentelemetry/instrumentation-graphql';
+import { PinoInstrumentation } from '@opentelemetry/instrumentation-pino';
+import { GrpcInstrumentation } from '@opentelemetry/instrumentation-grpc';
+import { TypeormInstrumentation } from 'opentelemetry-instrumentation-typeorm';
+
+import { NodeSDK } from '@opentelemetry/sdk-node';
+import {
+  SimpleSpanProcessor,
+  BatchSpanProcessor,
+} from '@opentelemetry/sdk-trace-node';
+import * as opentelemetry from '@opentelemetry/api';
+import type { Span, SpanOptions } from '@opentelemetry/api';
+import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
+import { TraceExporter } from '@google-cloud/opentelemetry-cloud-trace-exporter';
+import { CloudPropagator } from '@google-cloud/opentelemetry-cloud-trace-propagator';
+
+// import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-grpc';
+// import { PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
+
+import {
+  envDetectorSync,
+  hostDetectorSync,
+  osDetectorSync,
+  processDetectorSync,
+} from '@opentelemetry/resources';
+import { containerDetector } from '@opentelemetry/resource-detector-container';
+import { gcpDetector } from '@opentelemetry/resource-detector-gcp';
+
+import {
+  // SemanticResourceAttributes,
+  SemanticAttributes,
+} from '@opentelemetry/semantic-conventions';
+
+import {
+  ConsoleLogRecordExporter,
+  SimpleLogRecordProcessor,
+} from '@opentelemetry/sdk-logs';
+
+import { diag, DiagConsoleLogger } from '@opentelemetry/api';
+import type { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify';
+import dc from 'node:diagnostics_channel';
+import { Message } from '@google-cloud/pubsub';
+const channel = dc.channel('fastify.initialization');
+
+declare module 'fastify' {
+  interface FastifyRequest {
+    span: opentelemetry.Span;
+  }
+
+  interface FastifyInstance {
+    tracer: opentelemetry.Tracer;
+  }
+}
+
+// const metricExporter = new OTLPMetricExporter({
+//   // hostname: 'jaeger-collector',
+//   url: `http://${process.env.OTLP_COLLECTOR_HOST}:${process.env.OTLP_COLLECTOR_PORT}`,
+// });
+const isProd = process.env.NODE_ENV === 'production';
+const resourceDetectors = [
+  envDetectorSync,
+  hostDetectorSync,
+  osDetectorSync,
+  processDetectorSync,
+  containerDetector,
+  gcpDetector,
+];
+
+const traceExporter = isProd
+  ? new TraceExporter({
+      // /** option 1. provide a service account key json */
+      // keyFile: './service_account_key.json',
+      // keyFilename: './service_account_key.json',
+      //
+      // /** option 2. provide credentials directly */
+      // credentials: {
+      //   client_email: 'string',
+      //   private_key: 'string',
+      // },
```

Comment:
For production the pod is already associated with a service account so we shouldn't provide the secret directly. Just make sure it has permissions for tracing and metrics